File: core/src/main/kotlin/org/bitlap/core/sql/rel/BitlapProject.kt
package org.bitlap.core.sql.rel

import org.apache.calcite.plan.RelOptCluster
import org.apache.calcite.plan.RelTraitSet
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.core.Project
import org.apache.calcite.rel.hint.RelHint
import org.apache.calcite.rel.type.RelDataType
import org.apache.calcite.rel.type.RelDataTypeFactory
import org.apache.calcite.rex.RexInputRef
import org.apache.calcite.rex.RexNode
import org.apache.calcite.rex.RexShuttle

/**
 * Project logical plan, see [org.apache.calcite.rel.logical.LogicalProject]
 */
class BitlapProject(
    cluster: RelOptCluster,
    traitSet: RelTraitSet,
    hints: List<RelHint>,
    input: RelNode,
    projects: List<RexNode>,
    rowType: RelDataType,
    override var parent: RelNode? = null,
) : Project(cluster, traitSet, hints, input, projects, rowType), BitlapNode {

    override fun copy(
        traitSet: RelTraitSet,
        input: RelNode,
        projects: MutableList<RexNode>,
        rowType: RelDataType,
    ): BitlapProject {
        return BitlapProject(cluster, traitSet, hints, input, projects, rowType, parent)
    }

    override fun withHints(hintList: List<RelHint>): RelNode {
        return BitlapProject(cluster, traitSet, hintList, input, projects, getRowType())
    }

    override fun deepHashCode(): Int {
        return super.deepHashCode0()
    }

    override fun deepEquals(obj: Any?): Boolean {
        return super.deepEquals0(obj)
    }

    companion object {
        fun of(
            cluster: RelOptCluster,
            traitSet: RelTraitSet,
            hints: List<RelHint>,
            input: RelNode,
            projects: List<RexNode>,
            rowType: RelDataType,
            parent: RelNode? = null
        ): BitlapProject {
            val fixRowType = this.fixRowType(cluster.typeFactory, projects, rowType, input.rowType)
            return BitlapProject(cluster, traitSet, hints, input, projects, fixRowType, parent)
        }

        // fix any type
        fun fixRowType(
            typeFactory: RelDataTypeFactory,
            projects: List<RexNode>,
            currentRowType: RelDataType,
            inputRowType: RelDataType
        ): RelDataType {
            val builder = typeFactory.builder()
            currentRowType.fieldList.forEachIndexed { idx, rt ->
                val refIndex = mutableListOf<Int>()
                projects[idx].accept(object : RexShuttle() {
                    override fun visitInputRef(inputRef: RexInputRef): RexNode {
                        refIndex.add(inputRef.index)
                        return super.visitInputRef(inputRef)
                    }
                })
                if (rt.type.sqlTypeName.name.uppercase() == "ANY" && refIndex.isNotEmpty()) {
                    builder.add(inputRowType.fieldList[refIndex.first()])
                } else {
                    builder.add(rt)
                }
            }
            return builder.build()
        }
    }
}

File: bitlap-core-s/src/main/scala/org/bitlap/core/sql/rel/BitlapProject.scala
/** Copyright (C) 2023 bitlap.org .
 */
package org.bitlap.core.sql.rel

import java.util.{ Collections, List as JList, Set as JSet }

import org.apache.calcite.plan.RelOptCluster
import org.apache.calcite.plan.RelTraitSet
import org.apache.calcite.rel.`type`.RelDataType
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.core.{ CorrelationId, Project }
import org.apache.calcite.rel.hint.RelHint
import org.apache.calcite.rex.RexNode

/** Project logical plan, see [org.apache.calcite.rel.logical.LogicalProject]
 */
class BitlapProject(
  cluster: RelOptCluster,
  traitSet: RelTraitSet,
  hints: JList[RelHint],
  input: RelNode,
  projects: JList[RexNode],
  rowType: RelDataType,
  private val _variablesSet: JSet[CorrelationId] = Collections.emptySet(),
  var parent: RelNode = null)
    extends Project(cluster, traitSet, hints, input, projects, rowType, _variablesSet)
    with BitlapNode {

  def this(
    cluster: RelOptCluster,
    traitSet: RelTraitSet,
    hints: JList[RelHint],
    input: RelNode,
    projects: JList[RexNode],
    rowType: RelDataType,
    parent: RelNode
  ) = this(cluster, traitSet, hints, input, projects, rowType, Collections.emptySet(), parent)

  override def copy(
    traitSet: RelTraitSet,
    input: RelNode,
    projects: JList[RexNode],
    rowType: RelDataType
  ): BitlapProject = {
    return BitlapProject(getCluster, traitSet, getHints, input, projects, rowType, Collections.emptySet(), parent)
  }

  override def withHints(hintList: JList[RelHint]): RelNode = {
    return BitlapProject(
      getCluster,
      getTraitSet,
      hintList,
      getInput,
      getProjects,
      getRowType,
      Collections.emptySet(),
      null
    )
  }

  override def deepHashCode(): Int = {
    return super.deepHashCode0()
  }

  override def deepEquals(obj: Any): Boolean = {
    return super.deepEquals0(obj)
  }

  def copy(input: RelNode, rowType: RelDataType, projects: JList[RexNode]): BitlapProject = {
    return BitlapProject(getCluster, traitSet, getHints, input, projects, rowType, parent)
  }
}
